### ISSUE TITLE
Do not load the whole file while indexing metadata

### ISSUE THREAD

**Issue by allo-** [mopidy/mopidy#1360]
mopidy seems really slow when indexing files on a NFS share and from the traffic it makes it seems that it loads the whole file. Can it be reduced to loading the few bytes of the metadata header?

**Comment by adamcik** [comment]
Based on your question I've looked into and opened #1361 which might save us creating the decoders when we don't ever intend to decode the audio. But as for controlling how much of the file is read I don't think I have any simple way of doing this with GStreamer.

**Comment by allo-** [comment]
is it an option to use libid3  and the corresponding ogg/flac id-libraries for these types? This will cover the most common use cases.

Currently mopidy is way to slow to index my < 20 GB library, while MPD has no problem. I think most of these libraries have python bindings.

**Comment by adamcik** [comment]
No, we've chosen to use GStreamer so that we don't have to deal with per media type libraries. So this is not an option for us. We already take care to avoid going into the playing state to not read more than what we need and with #1361 we would further limit this I believe. Beyond that further fixes depend https://bugzilla.gnome.org/show_bug.cgi?id=564749 which I believe was meant to address part of what you are asking for. 

As for speed, the library I've used for bench-marking NFS / remote scanning performance has about 1.4T and 190k files and can be scanned in a bit over an hour. Though this is over a local network with decent networking gear and 1000Mb/s links ethernet between the client and server.

**Comment by allo-** [comment]
I'm using it on a raspberry pi, so the network throughput is not that fast. But MPD is fast enough, while i always gave up letting mopidy index everything and went with running mpd and mopidy at the moment. It's just way too slow. The linked thing seems to solve it, else could it be something, which can be done in a plugin? Or would this need to replace the whole local-backend?

**Comment by adamcik** [comment]
There is nothing stopping you from doing the `mopidy local scan` on the file server and then just using the db file remotely. That would probably be the easiest way to get this to work for you. Otherwise depending on the which local backend you use you "just" would need to create something that outputs compatible files.

As for the raspi, yes can see how that would be a problem speed wise. The gear I've been testing with is a lot faster. So not really a representative test for the raspi case, but it did help immensely in removing some previous bottle necks when looking for files to scan.

**Comment by allo-** [comment]
The Raspi Network isn't that great, for streaming files its okay, but loading 20 GB takes quite some time. I am not sure, but it even seems that CPU is more limiting than network. Not sure why it's that slow while indexing.

**Comment by adamcik** [comment]
An other option that I haven't retried in a while is having multiple scanners in parallel so that at least one of them can do work while the others are stuck waiting for IO. Though I'm unsure if this would be a wise choice on a PI.

As for reading the whole file, is that an assumption or something you've verified?

**Comment by allo-** [comment]
Its a assumtion. I measured until i stopped the indexing 600 MB network traffic caused by the indexer, which seems too much for just metadata. But as said the CPU was at 100% the whole time and the network could have transfered more than 600 MB during this time.

**Comment by adamcik** [comment]
I've added some instrumentation locally, and for a 5197844 byte file 4030880 bytes seem to get read. This seems to match what you've been seeing.

**Comment by adamcik** [comment]
Think I just found a way to get this to reading 65536 of 5197844 bytes basically we just need to put a buffer after the src with `max-size-buffers=1`. So reading 12% instead of 77%. This should probably help quite a bit.

**Comment by allo-** [comment]
Great! That's fast debugging :)

**Comment by adamcik** [comment]
I tried a bit more, last night and basically found that your millage may vary. I still had some files where it would read 4MB of 10MB no matter what tricks I tried. So we can probably put my change in even if it doesn't fix all the files. I just want to run it past some smoke tests.

**Comment by allo-** [comment]
By the way for anyone who want to buy a small device: mopidy runs A LOT smoother on a banana pi (1) than on a raspberry pi.

**Comment by adamcik** [comment]
I'll probably also setup my home raspi with nfs + a mopidy test env even though I don't use it for music just to get a better feel for this on a slow system.

**Comment by allo-** [comment]
Have a look at http://www.pimusicbox.com/

But they do something better than installing mopidy on a raspbian rPi, because it ran much smoother there (only tested spotify) than installing it myself.

Either the problem was some mix of hardfloat / softfloat libraries, or my other raspberry pi with some other stuff on it was just overloaded. Hard to say, CPU and RAM seemed not to be the bottleneck.
